### PR TITLE
[KUBOS-350] Adding a socket interface to CSP

### PR DESCRIPTION
**Pull request by plauche**
As part of [KUBOS-350](https://kubostech.atlassian.net/browse/KUBOS-350) CSP needs the ability to communicate with other processes running in a Linux system.

 - [x] Create basic CSP socket interface
 - [x] Allow for multiple client connections to server (not sure if this is done in CSP or at the next layer up)

### CODE REVIEW

File: libcsp/CMakeLists.txt [hunk 1]
```diff
@@ -11,13 +11,13 @@ option (HMAC "" OFF)
 option (XTEA "" OFF)
 option (BINDINGS "" OFF)
 option (EXAMPLES "" OFF)
-option (DEBUG "" OFF)
+option (DEDUP "" OFF)
```
Comment (plauche) [inline libcsp/CMakeLists.txt]: Turns out the DEDUP flag is not a typo haha but is actually used to control the inclusion of `csp_dedup.c`.

File: libcsp/CMakeLists.txt [hunk 2]
```diff
@@ -74,6 +74,7 @@ add_library (csp STATIC
     $<$<AND:$<BOOL:${TARGET_LIKE_KUBOS_RT}>,$<BOOL:${YOTTA_CFG_CSP_USART}>>:src/drivers/usart/usart_kubos.c>
     src/interfaces/csp_if_lo.c
     $<$<BOOL:${IF_KISS}>:src/interfaces/csp_if_kiss.c>
+    src/interfaces/csp_if_socket.c
```
Comment (plauche) [inline libcsp/CMakeLists.txt]: This should probably be guarded by a `$<$<BOOL:${IF_SOCKET}>`

File: libcsp/CMakeLists.txt [hunk 3]
```diff
@@ -11,13 +11,13 @@ option (HMAC "" OFF)
...
 option (EXAMPLES "" OFF)
-option (DEBUG "" OFF)
+option (DEDUP "" OFF)
 option (VERBOSE "" OFF)
 option (IF_I2C "" OFF)
 option (IF_KISS "" OFF)
 option (IF_CAN "" OFF)
 option (IF_ZMQHUB "" OFF)
-option (CSP_DEBUG "" OFF)
+option (CSP_DEBUG "" ON)
```
Comment (plauche) [inline libcsp/CMakeLists.txt]: Remember to turn this off

File: libcsp/src/interfaces/csp_if_socket.c [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        if (bind(server_socket, (struct sockaddr*)&server, sizeof(server)) < 0) {
+            csp_log_error("Failed to bind\n");
+            return CSP_ERR_DRIVER;
+        }
+
+        listen(server_socket, 3);
+        c = sizeof(struct sockaddr_in);
+
+        csp_log_info("Wait to accept\n");
+        csp_socket_handle = accept(server_socket, (struct sockaddr *)&client, (socklen_t*)&c);
```
Comment (plauche) [inline libcsp/src/interfaces/csp_if_socket.c]: Is it correct/wise to perform the entirety of the socket connection init right here?

File: libcsp/src/interfaces/csp_if_socket.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    csp_iflist_add(&csp_if_socket);
+
+    return CSP_ERR_NONE;
+}
+
+/** Interface definition **/
+csp_iface_t csp_if_socket = {
+    .name = "socket",
+    .nexthop = csp_socket_tx,
+    .mtu = 250
```
Comment (plauche) [inline libcsp/src/interfaces/csp_if_socket.c]: I think the `.mtu` should be hardcoded to the read buffer size (`BUF_SIZE`)?

**Comment by plauche** [comment]
@kubostech/devs this is more or less ready for :eyes: :)

File: libcsp/include/csp/drivers/socket.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ * @{
+ */
+
+#ifndef SOCKET_H_
+#define SOCKET_H_
+
+#include <csp/interfaces/csp_if_socket.h>
+
+/**
+ *
```
Comment (CatFreed) [inline libcsp/include/csp/drivers/socket.h]: nit: Missing function descriptions

File: libcsp/src/drivers/socket/socket_linux.c [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+int socket_status(const csp_socket_handle_t * socket_iface) {
+    int error;
+    int retval;
+    socklen_t len;
+
+    retval = getsockopt(socket_iface->socket_handle, SOL_SOCKET, SO_ERROR, &error, &len);
+    if ((retval != 0) || (error != 0)) {
+        return CSP_ERR_DRIVER;
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/socket_linux.c]: Might be useful to log the value of error if it's not zero

File: libcsp/src/drivers/socket/socket_linux.c [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    csp_log_info("Socket created");
+    
+    server.sin_addr.s_addr = inet_addr(addr);
+    server.sin_family = AF_INET;
+    server.sin_port = htons(port);
+
+    //Connect to remote server
+    if (connect(socket_handle, (struct sockaddr *)&server , sizeof(server)) < 0) {
+        csp_log_error("connect failed. Error");
```
Comment (CatFreed) [inline libcsp/src/drivers/socket/socket_linux.c]: nit: Capitalize connect

File: libcsp/src/interfaces/csp_if_socket.c [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+static int csp_socket_tx(csp_iface_t *ifc, csp_packet_t *packet, uint32_t timeout);
+
+
+/**
+ *
+ */
+CSP_DEFINE_TASK(csp_socket_rx);
+
+static int csp_socket_tx(csp_iface_t *ifc, csp_packet_t *packet, uint32_t timeout) {
+    csp_log_info("csp_socket_tx\r\n");
```
Comment (CatFreed) [inline libcsp/src/interfaces/csp_if_socket.c]: nit: This doesn't seem like a useful message.

File: libcsp/src/interfaces/csp_if_socket.c [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    if ((ifc == NULL) || (ifc->driver == NULL)) {
+        csp_log_error("Null pointer for interface or driver\r\n");
+        return CSP_ERR_DRIVER;
+    }
+
+    csp_socket_handle_t * socket_driver = ifc->driver;
+
+    csp_log_info("now we write\r\n");
```
Comment (CatFreed) [inline libcsp/src/interfaces/csp_if_socket.c]: nit: Also this one

File: libcsp/src/interfaces/csp_if_socket.c [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+    socket_iface->driver = socket_driver;
+    socket_iface->nexthop = csp_socket_tx;
+    socket_iface->name = "socket";
+    socket_iface->mtu = BUF_SIZE;
+
+    /* Start RX thread */
+	static csp_thread_handle_t handle_rx;
+	int ret = csp_thread_create(csp_socket_rx, "SOCKET_RX", 1000, socket_iface, 0, &handle_rx);
+	csp_log_info("Task start %d\r\n", ret);
```
Comment (CatFreed) [inline libcsp/src/interfaces/csp_if_socket.c]: Nit: spacing

**Pull request merged**